cudax/examples/cccl-rt/kernel_launch_patterns.cu

```diff
@@ -0,0 +1,217 @@
+__device__ void say_hello(dim3 tid, const name_buffer& name)
+{
+  printf("Hello world from thread [%u, %u] launched as %s!\n", tid.x, tid.y, name.data);
+
+  // Wait for all threads in block to print the output.
+  __syncthreads();
+
+  // Print additional new line once.
+  if (tid.x == 0 && tid.y == 0)
+  {
+    printf("\n");
+  }
+}
+
+// This is the traditional way to define a kernel, a void function decorated with __global__ attribute.
+__global__ void kernel(name_buffer name)
+{
+  say_hello(threadIdx, name);
+}
+
+// This is a kernel functor, a callable object with operator() decorated with __device__ attribute. When launched, the
+// object is copied to the device and operator() is invoked on the device.
+struct kernel_functor
+{
+  // The functor object can be set on host before the launch. Keep in mind that the functor (thus all the members) must
+  // be trivially copyable.
+  int member;
+
+  // The operator() must be decorated with __device__ attribute. It can also be a template.
+  __device__ void operator()(name_buffer name)
+  {
+    say_hello(threadIdx, name);
+
+    // Check that the member was copied correctly to the device.
+    assert(member == 42);
+  }
+};
+
+// This is again a kernel functor, but this time the operator() takes the implicit kernel configuration parameter. This
+// parameter is a cudax::kernel_config object that contains the launch configuration.
+struct kernel_functor_with_config
+{
+  // A type that represents the layout of the dynamic shared memory used by this kernel functor.
+  struct dynamic_smem_layout
+  {
+    int value;
+  };
+
+  // The operator() must be again decorated with __device__ attribute. It can also be a template or take additional
+  // parameters after the kernel configuration parameter. Return type must be void.
+  template <class Dims, class... Opts>
+  __device__ void operator()(cudax::kernel_config<Dims, Opts...> config, name_buffer name)
+  {
+    // dims.index(entity, in_level) queries the index of an entity in a hierarchy level. Query of the thread entity
+    // index in the block hierarchy level results in the same value as blockIdx.
+    const auto thread_idx = config.dims.index(cudax::thread, cudax::block);
+    assert(thread_idx.x == threadIdx.x);
+    assert(thread_idx.y == threadIdx.y);
+
+    say_hello(thread_idx, name);
+
+    // Similarly dims.extents(entity, level) queries the extents of an entity in a hierarchy level. Query of the thread
+    // entity extent in the block hierarchy results in the same value as blockDim.
+    const auto block_dim = config.dims.extents(cudax::thread, cudax::block);
+    assert(block_dim.x == blockDim.x);
+    assert(block_dim.y == blockDim.y);
+
+    // The main advantage of using kernel config instead of the ordinary kernel parameters is that the config can carry
+    // statically defined extents. That means it is easier to generate kernels specialized for certain block sizes.
+    static_assert(
```

Comment:
suggestion: I'd love to see an example that uses C++20 concepts/requires clause to statically constrain the `operator()` to a specific block size. I think this is where the ability to statically specify and query grid extents really shines.